imap/itip_support.c

```diff
@@ -0,0 +1,1053 @@
+#include "httpd.h"
+#include "ical_support.h"
+#include "jmap_ical.h"
+
+/* generated headers are not necessarily in current directory */
+#include "imap/imap_err.h"
+#include "imap/http_err.h"
+
+
+HIDDEN void sched_param_fini(struct caldav_sched_param *sparam)
+{
+    free(sparam->userid);
+    free(sparam->server);
+
+    struct proplist *prop, *next;
+    for (prop = sparam->props; prop; prop = next) {
+        next = prop->next;
+        free(prop);
+    }
+
+    memset(sparam, 0, sizeof(struct caldav_sched_param));
+}
+
+HIDDEN icalproperty *find_attendee(icalcomponent *comp, const char *match)
+{
+    if (!comp) return NULL;
+
+    icalproperty *prop = icalcomponent_get_first_invitee(comp);
+
+    for (; prop; prop = icalcomponent_get_next_invitee(comp)) {
+        const char *attendee = icalproperty_get_invitee(prop);
+        if (!attendee) continue;
+        if (!strncasecmp(attendee, "mailto:", 7)) attendee += 7;
+
+        /* Skip where not the server's responsibility */
+        icalparameter *param = icalproperty_get_scheduleagent_parameter(prop);
+        if (param) {
+            icalparameter_scheduleagent agent =
+                icalparameter_get_scheduleagent(param);
+            if (agent != ICAL_SCHEDULEAGENT_SERVER) continue;
+        }
+
+        if (!strcasecmp(attendee, match)) return prop;
+    }
+
+    return NULL;
+}
+
+HIDDEN const char *get_organizer(icalcomponent *comp)
+{
+    icalproperty *prop =
+        icalcomponent_get_first_property(comp, ICAL_ORGANIZER_PROPERTY);
+    const char *organizer = icalproperty_get_organizer(prop);
+    if (!organizer) return NULL;
+    if (!strncasecmp(organizer, "mailto:", 7)) organizer += 7;
+    icalparameter *param = icalproperty_get_scheduleagent_parameter(prop);
+    /* check if we're supposed to send replies to the organizer */
+    if (param &&
+        icalparameter_get_scheduleagent(param) != ICAL_SCHEDULEAGENT_SERVER)
+        return NULL;
+    return organizer;
+}
+
+static icalparameter_partstat get_partstat(icalcomponent *comp,
+                                           const char *attendee)
+{
+    icalproperty *prop = find_attendee(comp, attendee);
+    if (!prop) return ICAL_PARTSTAT_NEEDSACTION;
+    icalparameter *param =
+        icalproperty_get_first_parameter(prop, ICAL_PARTSTAT_PARAMETER);
+    if (!param) return ICAL_PARTSTAT_NEEDSACTION;
+    return icalparameter_get_partstat(param);
+}
+
+HIDDEN int partstat_changed(icalcomponent *oldcomp,
+                              icalcomponent *newcomp, const char *attendee)
+{
+    if (!attendee) return 1; // something weird is going on, treat it as a change
+    return (get_partstat(oldcomp, attendee) != get_partstat(newcomp, attendee));
+}
+
+/* Returns the calendar collection name to use as scheduling default.
+ * This is just the *last* part of the complete path without trailing
+ * path separator, e.g. 'Default' */
+HIDDEN char *caldav_scheddefault(const char *userid)
+{
+    const char *annotname =
```

Comment:
There's probably a ton of places where that change could be made throughout the code if we wanted.  We do use this pattern a lot.  It's probably not worth doing it one-off here though, because that just makes more different versions to look for if we decide on a different pattern than this.